awx_collection/plugins/module_utils/tower_api.py

```diff
@@ -143,63 +142,42 @@ def get_one(self, endpoint, name_or_id=None, *args, **kwargs):
             self.fail_json(msg="The endpoint did not provide count and results")
 
         if response['json']['count'] == 0:
-            return None
+            if allow_none:
+                return None
+            else:
+                self.fail_wanted_one(response, endpoint, new_kwargs.get('data'))
         elif response['json']['count'] > 1:
             if name_or_id:
                 # Since we did a name or ID search and got > 1 return something if the id matches
                 for asset in response['json']['results']:
-                    if asset['id'] == name_or_id:
+                    if str(asset['id']) == name_or_id:
                         return asset
             # We got > 1 and either didn't find something by ID (which means multiple names)
             # Or we weren't running with a or search and just got back too many to begin with.
-            self.fail_json(msg="An unexpected number of items was returned from the API ({0})".format(response['json']['count']))
+            self.fail_wanted_one(response, endpoint, new_kwargs.get('data'))
 
         return response['json']['results'][0]
 
-    def get_one_by_name_or_id(self, endpoint, name_or_id):
```

Comment:
This does almost the exact same thing as `get_one`, and it was relatively trivial to consolidate them.